csrc/optimization/alias_analysis.cpp

```diff
@@ -33,10 +33,52 @@ bool isContiguous(const TensorView& tv) {
   return true;
 }
 
-// Finds aliases of `root` and stores the findings in `alias_to_source`.
-void findAliasesOfRoot(
-    const TensorView* root,
-    AliasAnalysisResult& alias_to_source) {
+// Whether a ViewOp transforms any expanded broadcast IterDomain in the input.
+// This is a corner case in which we can't turn `out` into an alias.
+//
+// For example, given
+//
+//   t0 = makeContigConcreteTensor({4, 5});
+//   t1 = broadcast(t0, {false, false, true});
+//   t2 = expand(t1, {4, 5, 6});
+//
+// `reshape(t2, {40, 3})` and `reshape(t2, {4, 30})` have to copy data because
+// the former splits the expanded broadcast IterDomain (which is 6) and the
+// latter merges it.  However, `reshape(t2, {20, 6})` can simply be an alias
+// because the expanded broadcast IterDomain is forwarded not transformed.
+//
+// Obviously, this function assumes `in` and `out` are the input and output
+// TensorView of the same ViewOp.
+bool transformsExpandedBroadcastIterDomain(TensorView* in, TensorView* out) {
+  const std::vector<IterDomain*>& in_rfactor = in->getMaybeRFactorDomain();
+  const std::vector<IterDomain*>& out_root = out->getRootDomain();
+  const std::vector<IterDomain*>& out_rfactor = out->getMaybeRFactorDomain();
+
+  std::unordered_set<Val*> expanded_broadcast_dims;
+  for (size_t i = 0, size = in_rfactor.size(); i < size; i++) {
+    IterDomain* id = in_rfactor[i];
+    if (id->isBroadcast() && id->hasExpandedExtent()) {
+      expanded_broadcast_dims.insert(out_root[i]);
+    }
+  }
+
+  const std::vector<Expr*> transforms = DependencyCheck::getAllExprsBetween(
+      {out_root.begin(), out_root.end()},
+      {out_rfactor.begin(), out_rfactor.end()});
+
+  for (const auto* transform : transforms) {
```

Comment:
nitpick: can we put a comment here that we are being conservative here?

i.e. split expanded dimension should be fine I guess, as long as we don't merge it with other domains.
Meanwhile, I'm wondering if these analysis can be done `safer` through some replay. tagging @jacobhinkle for comments.

Reply:
> split expanded dimension should be fine I guess, as long as we don't merge it with other domains.

I agree with you, but the following test fails at the moment. I think nvFuser isn't smart enough yet to figure out the split expanded broadcast can remain stride=0. I'll file an enhancement... 

```
TEST_F(AliasAnalysisTest, View_SplitExpandedBroadcast) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  TensorView* in = makeContigConcreteTensor({4, 5});
  fusion.addInput(in);
  TensorView* out = broadcast(in, {false, false, true});
  out = expand(
      out,
      {IrBuilder::create<Val>(4),
       IrBuilder::create<Val>(5),
       IrBuilder::create<Val>(6)});
  out = reshape(out, {4, 5, 6}, {4, 5, 2, 3});
  fusion.addOutput(out);

  FusionExecutor fe;
  at::Tensor in_tensor =
      at::randn({4, 5}, at::dtype(at::kFloat).device(at::kCUDA, 0));
  fe.compileFusion(&fusion, {in_tensor});
  at::Tensor out_tensor = fe.runFusion({in_tensor})[0];
  EXPECT_THAT(out_tensor.strides(), ElementsAre(5, 1, 0, 0));
}
```

Reply:
Good catch!